### PR TITLE
[Audit] - Fetch Blocks Properly when Leader Consecutively

### PR DESCRIPTION
**Pull request by lukeiannucci**
Closes #[3462](https://github.com/EspressoSystems/HotShot/issues/3462)

### This PR: 

Fixes a bug where a node could be the leader for the current and next view. Currently we only propose block for current view.

### CODE REVIEW

File: crates/hotshot/src/traits/election/static_committee_leader_two_views.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+{
+    /// Clone the public key and corresponding stake table for current elected committee
+    fn committee_qc_stake_table(&self) -> Vec<PUBKEY::StakeTableEntry> {
+        self.committee_nodes_with_stake.clone()
+    }
+
+    #[cfg(not(any(
+        feature = "randomized-leader-election",
+        feature = "fixed-leader-election"
+    )))]
```
Comment (bfish713) [inline crates/hotshot/src/traits/election/static_committee_leader_two_views.rs:66]: I don't think it makes sense to have these features inside this impl.  I think this should just apply to the entire struct, or not have the feature gating at all

File: crates/hotshot/src/traits/election/static_committee_leader_two_views.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    fn leader(&self, view_number: TYPES::Time) -> PUBKEY {
+        // two connsecutive views will have same index starting with even number.
+        // eg 0->1, 2->3 ... 10->11 etc
+        let index =
+            usize::try_from((*view_number / 2) % self.all_nodes_with_stake.len() as u64).unwrap();
+        let res = self.all_nodes_with_stake[index].clone();
+        TYPES::SignatureKey::public_key(&res)
+    }
+
+    #[cfg(feature = "fixed-leader-election")]
```
Comment (bfish713) [inline crates/hotshot/src/traits/election/static_committee_leader_two_views.rs:77]: See above comment, but given this impl is specifically for a certain leader election we should not be implementing the fixed leader version

File: crates/hotshot/src/traits/election/static_committee_leader_two_views.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            || self.fixed_leader_for_gpuvid > self.all_nodes_with_stake.len()
+        {
+            panic!("fixed_leader_for_gpuvid is not set correctly.");
+        }
+        let index = usize::try_from(*view_number % self.fixed_leader_for_gpuvid as u64).unwrap();
+        let res = self.all_nodes_with_stake[index].clone();
+        TYPES::SignatureKey::public_key(&res)
+    }
+
+    #[cfg(feature = "randomized-leader-election")]
```
Comment (bfish713) [inline crates/hotshot/src/traits/election/static_committee_leader_two_views.rs:91]: Same as above

File: crates/task-impls/src/transactions.rs [hunk 4]
```diff
@@ -360,30 +360,41 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> TransactionTaskState<TYPES,
...
+                let block_views = match (make_block, next_leader) {
+                    // case where we are next leader and current leader
+                    (true, true) => vec![self.cur_view, next_view],
+                    // current view leader
+                    (true, false) => vec![self.cur_view],
+                    // next view leader
+                    (false, true) => vec![next_view],
+                    (false, false) => {
+                        // return if we aren't the next leader or we skipped last view and aren't the current leader.
+                        debug!("Not next leader for view {:?}", self.cur_view);
```
Comment (bfish713) [inline crates/task-impls/src/transactions.rs:374]: I'm not sure if it's better but this logic, and the for loop below could be 2 if statements:

```rust
if make_block {
    // make block for self.cur_view
}
if next_leader {
    // make block for next_view
}
```

At the very least I think we can get rid of this match statement and push the just like above

File: crates/task-impls/src/transactions.rs [hunk 5]
```diff
@@ -360,30 +360,41 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> TransactionTaskState<TYPES,
...
+                        }
+                    };
+
+                    if version < MarketplaceVersion::VERSION {
+                        self.handle_view_change_legacy(&event_stream, block_view)
+                            .await;
+                    } else {
+                        self.handle_view_change_marketplace(&event_stream, block_view)
+                            .await;
+                    }
```
Comment (bfish713) [inline crates/task-impls/src/transactions.rs:397]: If you switch to 2 if statments maybe we can refactor this into a simple helper like `handle_view_change(&self, event_stream: &Sender<..>, view: TYPES::Time)`

**Review by bfish713 — COMMENTED** [review]
Overall looks good, just had a few suggestions to remove code, and maybe make the logic in the transaction task a bit easier to follow, but that's up to you

File: crates/task-impls/src/transactions.rs (hunk 5 above)
Comment (lukeiannucci) [reply crates/task-impls/src/transactions.rs:397]: agreed, this was cleaned up recently which i think switch to the above idea works better now.

File: crates/hotshot/src/traits/election/static_committee_leader_two_views.rs (hunk 1 above)
Comment (lukeiannucci) [reply crates/hotshot/src/traits/election/static_committee_leader_two_views.rs:66]: yep i had it before i did some refactoring and making the test build_system_handle generic. now it is not needed

**Review by bfish713 — APPROVED** [review]

**Pull request merged**